ARC: Add multcore (in SMP arrangement) ARC HS in nSIM

abrodkin opened the pull request:
**NOTE:** There's a prerequisite - https://github.com/zephyrproject-rtos/zephyr/pull/17747 and so only the last commit in this fake series to be reviewed & discussed as after https://github.com/zephyrproject-rtos/zephyr/pull/17747 is merged (hopefully anytime now) I'll rebase on top of it and those commits will be gone.

Now when SMP support for ARC is available we may introduce a simulation platform which might be used for testing & app development for SMP setups.

One important note is stand-alone nSIM (as well as its "Free" flavour) doesn't support SMP simulation so we have to switch to use of nSIM via proprietary MetaWare debugger (https://www.synopsys.com/dw/ipdir.php?ds=sw_metaware) and so:
 1. We introduce new emulation target "mdb"
 2. It's only possible to run that platform for those who have MetaWare tools installed and valid license.

Though QEMU port for ARC is in work at the moment and once we open that port and it has SMP support we'll switch to it and everybody will be able to try ARC HS with SMP.

If one has MDB license to try how SMP works just:
```
# mkdir build
# cd build
# cmake -DBOARD=nsim_hs_smp ../tests/kernel/smp/
# make run

System Architecture Configuration
--------------------------------------------------------------------------------
+ SYSTEM:'__TOP__' { RAM: [ latency:5, clock_freq:250[MHz], clock_divider:1, bus_width:32 ] }
  + MODULE:'MOD0'
    + CONNECT:'MOD0_CONNECT' {
          V:2 - ASI:0, ICI:2, ICS:1, ICM:1, PMU:1, ICD:2, GFRC:1, LLM:0, IDU:2, PDM:0
        , aap_num:0, ics_num_semas:16, pmu_dvfs:0, icm_sram_size:512, icm_sram_ecc:0, ici_irq:19
        , llm_size:131072, llm_base:2, llm_ecc:0, idu_cirqnum:64, idu_irq_base:24
      }
    + SCU:'SHARED COHERENCY UNIT'
    + CORE:'MOD0_CPU0' { UARCH:HS }
      + CACHE:'CPU0_ICACHE' { size:64KB, line:64B, sets:256, ways:4, repl:Pseudo-Random }
      + CACHE:'CPU0_DCACHE' { size:64KB, line:64B, sets:512, ways:2, repl:Pseudo-Random }
      + BPU:'CPU0_BPU' { HS - ways:1, sets:2048, pt_entries:16384, ras_entries:8 }
    + CORE:'MOD0_CPU1' { UARCH:HS }
      + CACHE:'CPU1_ICACHE' { size:64KB, line:64B, sets:256, ways:4, repl:Pseudo-Random }
      + CACHE:'CPU1_DCACHE' { size:64KB, line:64B, sets:512, ways:2, repl:Pseudo-Random }
      + BPU:'CPU1_BPU' { HS - ways:1, sets:2048, pt_entries:16384, ras_entries:8 }
--------------------------------------------------------------------------------
Console now belongs to UART, hit CRTL-] to return to simulator.
***** Booting Zephyr OS build zephyr-v1.14.0-2835-g36f26ad68ab5 *****
Running test suite smp
===================================================================
starting test - test_smp_coop_threads
PASS - test_smp_coop_threads
===================================================================
starting test - test_cpu_id_threads
PASS - test_cpu_id_threads
===================================================================
starting test - test_coop_resched_threads
PASS - test_coop_resched_threads
===================================================================
starting test - test_preempt_resched_threads
PASS - test_preempt_resched_threads
===================================================================
starting test - test_yield_threads
PASS - test_yield_threads
===================================================================
starting test - test_sleep_threads
PASS - test_sleep_threads
===================================================================
starting test - test_wakeup_threads
PASS - test_wakeup_threads
===================================================================
Test suite smp succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

abrodkin committed:
board/nsim: Add support of multi-core ARC HS platform in nSIM

Now when SMP support for ARC is available we may introduce a simulation
platform which might be used for testing & development for SMP setups.

One important note is stand-alone nSIM (as well as its "Free" flavour)
doesn't support SMP simulation so we have to switch to use of nSIM via
proprietary MetaWare debugger [1] and so:
 1. We introduce new emulation target "mdb"
 2. It's only possible to run that platform for those who
    have MetaWare tools installed and valid license.

Though QEMU port for ARC is in work at the moment and once we
open that port and it has SMP support we'll switch to it and everybody
will be able to try ARC HS with SMP.

[1] https://www.synopsys.com/dw/ipdir.php?ds=sw_metaware

Signed-off-by: Alexey Brodkin <abrodkin@synopsys.com>

abrodkin commented:
Re-based on the current master so that only one patch that matters is left.
Should be ready for merge.

ruuddw commented on boards/arc/nsim/nsim_hs_smp_defconfig:
```diff
@@ -0,0 +1,18 @@
+
+CONFIG_ARC=y
+CONFIG_CPU_ARCHS=y
+CONFIG_SOC_NSIM=y
```
Do you need both SOC_NSIM and SOC_NSIM_SMP set to yes?
Seems odd, having two SoCs selected for a single board.

abrodkin replied:
Well, this is how we used to handle different flavors of nSIM boards: we have `SOC_NSIM` for this platform in general and then `SOC_NSIM_EM`, `SOC_NSIM_SEM`, `SOC_NSIM_HS` & now `SOC_NSIM_HS_SMP`.

If I remove `SOC_NSIM` from `boards/arc/nsim/nsim_hs_smp_defconfig` on configuration I'm getting:
> warning: the choice symbol SOC_NSIM_HS_SMP (defined at soc/arc/snps_nsim/Kconfig:28) was selected (set =y), but no symbol ended up as the choice selection. 

and the whole lot of misconfiguration messages.

So I would agree that `CONFIG_SOC_NSIM` might be implicitly selected by `CONFIG_SOC_NSIM_HS_SMP` but that will require rework of nSIM Kconfig infrastructure.

That said IMHO it's better to merge nSIM HS_SMP changes first and then do a clean-up.

ruuddw replied:
Let's leave it like this, indeed for other ARC boards this is the same.

abrodkin commented:
@ruuddw could you please look if my comment addresses your concern?

ruuddw commented:
Shippable complaints, looks like openocd Zephyr awareness doesn't work with SMP (tries to get offset of the 'current' field in struct z_kernel, but that doesn't exist in SMP.
It's not really related to this patch, but would be good to fix. Maybe disable openocd debugging support for now, and create a new issue?

abrodkin commented:
@ruuddw indeed I completely forgot to run sanity check for `nsim_hs_smp` and treated Shippable build failure as yet another false positive one (seeing mention of "openocd" which looked completely unrelataed to me).

Reported https://github.com/zephyrproject-rtos/zephyr/issues/18124 and fixed with https://github.com/zephyrproject-rtos/zephyr/pull/18125.

So once https://github.com/zephyrproject-rtos/zephyr/pull/18125 gets merged we may rebase this one or merge this one first if that's OK as now sanity checker is happy:
```
./scripts/sanitycheck -p nsim_hs_smp -b
JOBS: 96
Cleaning output directory .../zephyr/sanity-out
Building testcase defconfigs...
Filtering test cases...
140 tests selected, 119003 tests discarded due to filters
total complete:  138/ 138  100%  failed:    0
138 of 138 tests passed with 0 warnings in 112 seconds
```

ruuddw commented:
Thanks for trying to do a quick fix for openocd - not sure though if that is the right fix and gets merged soon. Why not disabling the (experimental) OPENOCD_SUPPORT config for this target to not be dependent?

dbkinder reviewed (approved):
+1 for doc changes

abrodkin commented:
@nashif @andrewboie any thoughts on what should be done here? Do we need to fix https://github.com/zephyrproject-rtos/zephyr/issues/18124 first given it's a generic problem and should have been caught long ago on either SMP platform?

andrewboie commented:
> @nashif @andrewboie any thoughts on what should be done here? Do we need to fix #18124 first given it's a generic problem and should have been caught long ago on either SMP platform?

IMO put a workaround in this PR to disable the test. And leave the bug open so someone still fixes it, with a note saying to revert the workaround once fixed.

abrodkin commented:
> 
> 
> > @nashif @andrewboie any thoughts on what should be done here? Do we need to fix #18124 first given it's a generic problem and should have been caught long ago on either SMP platform?
> 
> IMO put a workaround in this PR to disable the test. And leave the bug open so someone still fixes it, with a note saying to revert the workaround once fixed.

You mean disable `OPENOCD_SUPPORT` for this particular platform?

andrewboie commented:
> You mean disable `OPENOCD_SUPPORT` for this particular platform?


I mean, do whatever is least invasive to get CI to stop rejecting this PR.

abrodkin commented:
> 
> 
> > You mean disable `OPENOCD_SUPPORT` for this particular platform?
> 
> I mean, do whatever is least invasive to get CI to stop rejecting this PR.

I'm not sure in Kconfig it's possible to black-list an option from a board's defconfig, so if I'm not missing anything we may add some dependency for `OPENOCD_SUPPORT`.

Either:
```
diff --git a/subsys/debug/Kconfig b/subsys/debug/Kconfig
index a55d0f3350..a5dff0ad4f 100644
--- a/subsys/debug/Kconfig
+++ b/subsys/debug/Kconfig
@@ -223,6 +223,7 @@ config EXCEPTION_STACK_TRACE

 config OPENOCD_SUPPORT
        bool "OpenOCD support [EXPERIMENTAL]"
+       depends on !SMP
        select THREAD_MONITOR
        select THREAD_NAME
        help
```
or
```
diff --git a/subsys/debug/Kconfig b/subsys/debug/Kconfig
index a55d0f3350..a5dff0ad4f 100644
--- a/subsys/debug/Kconfig
+++ b/subsys/debug/Kconfig
@@ -223,6 +223,7 @@ config EXCEPTION_STACK_TRACE

 config OPENOCD_SUPPORT
        bool "OpenOCD support [EXPERIMENTAL]"
+       depends on !SOCNSIM_HS_SMP
        select THREAD_MONITOR
        select THREAD_NAME
        help
```

latter is only applicable to this particular PR but makes not that much sense as the problem is for all SMP platforms, former is IMHO better but has nothing to do with this particular PR, thought I may submit it as a temporary work-around for https://github.com/zephyrproject-rtos/zephyr/issues/18124.

ruuddw commented:
I don't think anything is required for this PR. First, I thought OPENOCD_SUPPORT was explicitly enabled somewhere in this PR. However, it isn't. The shippable fail was unrelated, just a random config that happened to fail (zephyr/samples/synchronization/sample.yaml enables it - now triggered by ARC HS SMP, but could be triggered by any SMP target).

ruuddw reviewed (approved):
See comments. The shippable failure is unrelated and can't be fixed in this PR. I believe this can be merged.

ioannisg commented:
@vonhust can you check the CI error?

Commit:
arch: arc: fix a bug when CONFIG_SMP is enabled

the bug is forgot to fixes, when CONFIG_SMP is enabled,
it will cause build error

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
tests: enable hsdk and nsim_hs_smp for smp test

ARC now supports SMP, enable the kernel/smp test
for arc platforms

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

vonhust commented:
> @vonhust can you check the CI error?

I made some fixes, except the following one.

> I don't think anything is required for this PR. First, I thought OPENOCD_SUPPORT was explicitly enabled somewhere in this PR. However, it isn't. The shippable fail was unrelated, just a random config that happened to fail (zephyr/samples/synchronization/sample.yaml enables it - now triggered by ARC HS SMP, but could be triggered by any SMP target).

So it should be ok to merge

ioannisg commented:
> 
> 
> > @vonhust can you check the CI error?
> 
> I made some fixes, except the following one.
> 
> > I don't think anything is required for this PR. First, I thought OPENOCD_SUPPORT was explicitly enabled somewhere in this PR. However, it isn't. The shippable fail was unrelated, just a random config that happened to fail (zephyr/samples/synchronization/sample.yaml enables it - now triggered by ARC HS SMP, but could be triggered by any SMP target).
> 
> So it should be ok to merge

@vonhust thanks for the update, let's wait for CI. However, your fix-bug commit needs to be (formally) approved before we can merge that PR; is it possible for @ruuddw @abrodkin to give a +1 today? If not, I would move this to 2.1.0 release, is this OK with you?

vonhust commented:
> > > @vonhust can you check the CI error?
> > 
> > 
> > I made some fixes, except the following one.
> > > I don't think anything is required for this PR. First, I thought OPENOCD_SUPPORT was explicitly enabled somewhere in this PR. However, it isn't. The shippable fail was unrelated, just a random config that happened to fail (zephyr/samples/synchronization/sample.yaml enables it - now triggered by ARC HS SMP, but could be triggered by any SMP target).
> > 
> > 
> > So it should be ok to merge
> 
> @vonhust thanks for the update, let's wait for CI. However, your fix-bug commit needs to be (formally) approved before we can merge that PR; is it possible for @ruuddw @abrodkin to give a +1 today? If not, I would move this to 2.1.0 release, is this OK with you?

+1 from @ruudw @abrodkin may be a litte difficult as it's weekend now.  

My two commits are not only for this PR,  but also a generci fix， so i still hope this PR can be merged for 2.0. 

Now there is a still failure related to OPENOCD support.  the correct fix is #18125 .

I can make a workaround to let PR pass CI, leave #18125 to trace this issue

Commit:
boards: nsim: disable CONFIG_SMP by default for nsim_hs_smp

* CONFIG_SMP can be enabled by smp application, e.g. tests/kernel/smp
* if application is not designed for smp, CONFIG_SMP can be disabled,
and the target works as a single processor.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

ioannisg commented:
@vonhust your last commits are bug fix patches and can be merged for 2.0 during stabilization period, no problem. But the board support is not, so it needs to go in now or wait for 2.1. Now, I can't really review the fault.S patch myself, need someone with Arc knowledge

ioannisg commented on arch/arc/core/fault_s.S:
```diff
@@ -302,9 +302,10 @@ exc_nest_handle:
 	/* note: Ok to use _CAUSE_RIRQ since everything is saved */
 
-	ld_s r2, [r1, _kernel_offset_to_ready_q_cache]
+	mov r2, r0
```
We merge this one, but this patch, here, needs to be reviewed, @abrodkin @ruuddw @vonhust. Please, submit a fix if required.

ruuddw replied:
Thanks @Ioannisg for merging this one with the deadline of 2.0 merge window closing!
Btw, I already reviewed and approved this PR (including analysis and a remark on the unrelated CI failure).
What's your concern about this patch?

ruuddw replied:
I see this is an update Wayne pushed as part of this PR. Looks ok to me. (I think the issue wasn't in the original PR but due to some other PRs/rebasing).

ioannisg replied:
@ruuddw no, this particular change was in a commit that was pushed after you approved, as far as I could see.

ioannisg reviewed (commented):
Merging this for 2.0 release. Made a comment for reviewing a submitted patch and fix if required, with a bug-fix PR.
